src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,147 @@
+     */
+    def generateScopeTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[Scope] = {
+      node match {
+        case w: Ast if w.getName == AstNodeName.Workflow =>
+          val workflowChildren = getChildrenList(w) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter) }
+          val workflow: Workflow = Workflow(w, wdlSyntaxErrorFormatter)
+          workflow.setChildren(workflowChildren)
+          Some(workflow)
+
+        case c: Ast if c.getName == AstNodeName.Call => Some(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter))
+
+        case s: Ast if s.getName == AstNodeName.Scatter =>
+          // Need to create the Scatter before generating its children to have the correct index layout order
+          val scatter = Scatter(s, scatterCount)
+          scatterCount += 1
+          val scatterChildren = getChildrenList(s) flatMap { generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter) }
+          scatter.setChildren(scatterChildren)
+          Some(scatter)
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+    }
+  }
+
+  /**
+   * Generate a Scope Tree from an Ast Node.
+   * Node must match a Scope (currently: Workflow, Scatter or Scope)
+   * @throws UnsupportedOperationException if node does not match any supported Scopes
+   * @return generated Scope Tree.
+   */
+  def generateTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Scope = {
+    new ScopeTreeGenerator().generateScopeTree(node, namespaces, tasks, wdlSyntaxErrorFormatter)
+      .getOrElse(throw new UnsupportedOperationException("Input node cannot be instantiated as a Scope"))
+  }
+
+  /**
+   * Helper method to collect Calls from a Seq of Scopes. 
+   * /!\ NON-RECURSIVE.
+   * @param scopes scopes to loop through
+   * @return Scopes instances that are Calls
+   */
+  def collectCalls(scopes: Seq[Scope]): Seq[Call] = scopes collect { case s: Call => s }
+
+  /**
+   * Recursively collect all Calls from the given scope.
+   * @param scope scope to gather Calls from
+   * @return all Calls inside the scope
+   */
+  def collectAllCallsRec(scope: Scope): Seq[Call] = {
+    collectCalls(Seq(scope)) ++ (scope.children flatMap collectAllCallsRec)
+  }
+
+  /**
+   * Helper method to collect Scatters from a Seq of Scopes. 
+   * /!\ NON-RECURSIVE.
+   * @param scopes scopes to loop through
+   * @return Scopes instances that are Scatters
+   */
+  def collectScatters(scopes: Seq[Scope]): Seq[Scatter] = scopes collect { case s: Scatter => s }
+
+  /**
+   * Recursively collect all Scatters from the given scope.
+   * @param scope scope to gather Scatters from
+   * @return all Scatters inside the scope
+   */
+  def collectAllScattersRec(scope: Scope): Seq[Scatter] = {
```

Comment:
Are the `Rec` portions of these names to note recursiveness? I don't think that's necessary. Although if you can fiddle them to be tail recursive you can use the annotation